Mathlib/Algebra/Order/Hom/Basic.lean

```diff
@@ -118,3 +139,140 @@ theorem le_map_div_add_map_div [Group α] [AddCommSemigroup β] [LE β] [MulLEAd
+
+@[to_additive] lemma le_map_add_map_div' : f x ≤ f y + f (y / x) :=
+by simpa only [add_comm, map_div_rev, div_mul_cancel'] using map_mul_le_add f (x / y) y
+
+end GroupSeminormClass
+
+example [OrderedAddCommGroup β] : OrderedAddCommMonoid β := inferInstance
+
+@[to_additive] lemma abs_sub_map_le_div [Group α] [LinearOrderedAddCommGroup β]
+  [GroupSeminormClass F α β] (f : F) (x y : α) : |f x - f y| ≤ f (x / y) := by
+  rw [abs_sub_le_iff, sub_le_iff_le_add', sub_le_iff_le_add']
+  exact ⟨le_map_add_map_div _ _ _, le_map_add_map_div' _ _ _⟩
+
+@[to_additive] -- See note [lower instance priority]
+instance (priority := 100) GroupSeminormClass.to_NonNegHomClass [Group α]
+  [LinearOrderedAddCommMonoid β] [GroupSeminormClass F α β] :
+  NonNegHomClass F α β :=
+{ ‹GroupSeminormClass F α β› with
+  map_nonneg := fun f a ↦ (nsmul_nonneg_iff two_ne_zero).1 $
+    by rw [two_nsmul, ←map_one_eq_zero f, ←div_self' a]; exact map_div_le_add _ _ _ }
+
+section GroupNormClass
+variable [Group α] [OrderedAddCommMonoid β] [GroupNormClass F α β] (f : F) {x : α}
+
+@[simp, to_additive] lemma map_eq_zero_iff_eq_one : f x = 0 ↔ x = 1 :=
+⟨eq_one_of_map_eq_zero _, by rintro rfl; exact map_one_eq_zero _⟩
+
+@[to_additive] lemma map_ne_zero_iff_ne_one : f x ≠ 0 ↔ x ≠ 1 := (map_eq_zero_iff_eq_one _).not
+
+end GroupNormClass
+
+@[to_additive] lemma map_pos_of_ne_one [Group α] [LinearOrderedAddCommMonoid β]
+  [GroupNormClass F α β] (f : F) {x : α} (hx : x ≠ 1) : 0 < f x :=
+(map_nonneg _ _).lt_of_ne $ ((map_ne_zero_iff_ne_one _).2 hx).symm
+
+/-! ### Ring (semi)norms -/
+
+/-- `RingSeminormClass F α` states that `F` is a type of `β`-valued seminorms on the ring `α`.
+
+You should extend this class when you extend `ring_seminorm`. -/
+class RingSeminormClass (F : Type _) (α β : outParam $ Type _) [NonUnitalNonAssocRing α]
+  [OrderedSemiring β] extends AddGroupSeminormClass F α β, SubMultiplicativeHomClass F α β
+
+/-- `RingNormClass F α` states that `F` is a type of `β`-valued norms on the ring `α`.
+
+You should extend this class when you extend `ring_norm`. -/
+class RingNormClass (F : Type _) (α β : outParam $ Type _) [NonUnitalNonAssocRing α]
+  [OrderedSemiring β] extends RingSeminormClass F α β, AddGroupNormClass F α β
+
+/-- `MulRingSeminormClass F α` states that `F` is a type of `β`-valued multiplicative seminorms
+on the ring `α`.
+
+You should extend this class when you extend `mul_ring_seminorm`. -/
+class MulRingSeminormClass (F : Type _) (α β : outParam $ Type _) [NonAssocRing α]
+  [OrderedSemiring β] extends AddGroupSeminormClass F α β, MonoidWithZeroHomClass F α β
+
+/-- `MulRingNormClass F α` states that `F` is a type of `β`-valued multiplicative norms on the
+ring `α`.
+
+You should extend this class when you extend `mul_ring_norm`. -/
+class MulRingNormClass (F : Type _) (α β : outParam $ Type _) [NonAssocRing α]
+  [OrderedSemiring β] extends MulRingSeminormClass F α β, AddGroupNormClass F α β
+
+-- TODO: Somehow this does not get inferred.
```

Comment:
Could you explain what this means specifically? Do you have an example of this instance not applying when it should?

Reply:
I get the same problem as in Lean 3. The following fails:
```
instance (priority := 100) RingSeminormClass.to_NonnegHomClass [NonUnitalNonAssocRing α]
  [LinearOrderedSemiring β] [RingSeminormClass F α β] : NonnegHomClass F α β :=
inferInstance
```

Reply:
```suggestion
-- This instance is simply the composition of the instances
-- `AddGroupSeminormClass.toNonnegHomClass` and `RingSeminormClass.toAddGroupSeminormClass`.
-- However, it has to be declared explicitly due to hitting an edge case in the inference algorithm:
-- those instances both depend on a different subclass of `OrderedAddMonoid β`, but `β` is an
-- `out_param` that cannot be inferred until both instances are applied.
-- So we have to do it manually until Lean becomes smarter.
```